src/ripple/app/paths/impl/DirectStep.cpp

```diff
@@ -821,24 +823,49 @@ DirectStepI<TDerived>::lineQualityIn (ReadView const& v) const
 }
 
 template <class TDerived>
-boost::optional<Quality>
-DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
+std::pair<boost::optional<Quality>, DebtDirection>
+DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection prevStepDir)
     const
 {
-    auto const prevStepDebtDir = dir;
-    dir = this->debtDirection(v, StrandDirection::forward);
-    std::uint32_t const srcQOut = [&]() -> std::uint32_t {
-        if (redeems(prevStepDebtDir) && issues(dir))
-            return transferRate(v, src_).value;
-        return QUALITY_ONE;
+    auto const dir = this->debtDirection(v, StrandDirection::forward);
+
+    if (!v.rules().enabled(fixQualityUpperBound))
+    {
+        std::uint32_t const srcQOut = [&]() -> std::uint32_t {
+            if (redeems(prevStepDir) && issues(dir))
+                return transferRate(v, src_).value;
+            return QUALITY_ONE;
+        }();
+        auto dstQIn = static_cast<TDerived const*>(this)->quality(
+            v, QualityDirection::in);
+
+        if (isLast_ && dstQIn > QUALITY_ONE)
+            dstQIn = QUALITY_ONE;
+        Issue const iss{currency_, src_};
+        return {Quality(getRate(STAmount(iss, srcQOut), STAmount(iss, dstQIn))),
+                dir};
+    }
+
+    auto const [srcQOut, dstQIn] = [&] {
+        if (dir == DebtDirection::redeems)
+        {
+            return qualitiesSrcRedeems(v);
+        }
+        else
+        {
+            return qualitiesSrcIssues(v, prevStepDir);
+        }
```

Comment:
Can this block become the following?

```cpp
auto const [srcQOut, dstQIn] = redeems(dir) ? qualitiesSrcRedeems(v) : qualitiesSrcIssues(v, prevStepDir);
```

Reply:
Yes, that's nicer. Fixed in `2847bf988 [fold] Replace lambda with conditional operator
`